File: codesignal/decodeString.js
/*
Given an encoded string, return its corresponding decoded string.

The encoding rule is: k[encoded_string], where the encoded_string inside the square brackets is repeated exactly k times. Note: k is guaranteed 
to be a positive integer.

Note that your solution should have linear complexity because this is what you will be asked during an interview.

Example

For s = "4[ab]", the output should be
decodeString(s) = "abababab";

For s = "2[b3[a]]", the output should be
decodeString(s) = "baaabaaa";

For s = "z1[y]zzz2[abc]", the output should be
decodeString(s) = "zyzzzabcabc".
*/

// first pass solution 
function decodeString(s) {
    let stack = [];
    
    for (let char of s) {
        if (char === ']') {
            let number = '';
            let curr = '';
            
            while (/[a-z]/.test(stack[stack.length - 1])) {
                curr = stack.pop() + curr;
            }
            stack.pop(); // pop off '['
            
            while (/[0-9]/.test(stack[stack.length - 1])) {
                number = stack.pop() + number;
            }
            stack.push(curr.repeat(number));
        }
        else {
            stack.push(char);
        }
    }
    return stack.join('');
}

File: leetcode/medium/decodeString.java
class Solution {
    public String decodeString(String s) {
        Stack<String> stack = new Stack<>();

        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
          
            if (ch != ']') {
                stack.push(String.valueOf(ch));
            } else {
                String chars = "";
                String nums = "";

                while (stack.size() > 0 && Character.isLetter(stack.peek().charAt(0))) {
                    chars = stack.pop() + chars;
                }

                if (stack.size() > 0 && stack.peek().charAt(0) == '[') {
                    stack.pop();
                }
              
                while (stack.size() > 0 && Character.isDigit(stack.peek().charAt(0))) {
                    nums = stack.pop() + nums;
                }
              
                if (nums.length() > 0) {
                    stack.push(chars.repeat(Integer.parseInt(nums)));
                } else {
                    stack.push(chars);
                }
            }
        }
        return String.join("", stack);
    }
}
